### PR TITLE
be able to pass additional arguments to F and G

### PR DESCRIPTION
**Pull request by lucidrains**
Re-opening with some fixes

---

This is probably the last PR to this repository I need. So I am trying to make the Reformer work as a decoder, where it accepts encoded keys from the encoder and does its work. However, as it is now, the reversible blocks accept only one argument, from which it does the splitting, residuals, and catting. I'd like to be able to pass all reversible modules in a reversible sequence a constant set of arguments for F and G.

Let me know if this looks right to you, or if you can think of a cleaner way to achieve this!

### CODE REVIEW

**Comment by RobinBruegger** [comment]
Hi @lucidrains 

I do not think you need a any change in RevTorch for what you try to achieve. `f_func` and `g_func` inherit from `nn.Module`, so you could do something like this.

```
class FFunc(nn.Module):
    def __init__(self, params_that_you_need):
        super(FFunc, self).__init__()
        self.params = params_that_you_need

    def forward(self, x):
        #use params here
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))
```

I have not written any Python in a while, so it might be somewhat off syntactically. Does this work for your usecase?

Btw, I think it would be possible to fix the random seed with a similar mechanism ;)

**Comment by lucidrains** [comment]
@RobinBruegger That could certainly work on my end! I think presenting a solution for others would make your library even more flexible and powerful, but I'll leave it up to you! :)

**Comment by RobinBruegger** [comment]
@lucidrains Thanks for the PR, I appreaciate it. I would like to strike a balance between adding features and keeping the library neat and easy to understand. Therefore I prefer not to add this feature, especially since the desired behaviour can be achieved with the mechanisms already in place ;)

**Comment by lucidrains** [comment]
Ok no probs! I'll go with the solution you outlined. Thanks for the great lib!

**Comment by lucidrains** [comment]
Oh no, I'm afraid that doesn't solve the problem.

I need to be able to pass the additional arguments at runtime to the decoder, so unless if I reach into the ReversibleSequence and set the params variable for each module, it would not really work. I'll continue with a hacked version where I set and deset the params variable before and after the forward, but if you can think of any solution, let me know. Thanks!

**Comment by lucidrains** [comment]
It's ok, I'll just work my way around this :)

**Comment by RobinBruegger** [comment]
You could create a class that holds the params and modify them from the outside. Hacky, but should work. I'll think about this again tomorrow evening

**Comment by lucidrains** [comment]
https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L455 Yup, that's what I did! :)
